### PR TITLE
Remove OneVector type alias

### PR DESCRIPTION
**Pull request by ljedrz**
Removes the `OneVector` type alias (equivalent to `SmallVec<[T; 1]>`); it is used in scenarios where the capacity of 1 is often exceeded, which might be nullifying the performance wins (due to spilling to the heap) expected when using `SmallVec` instead of `Vec`.

The numbers I used in this PR are very rough estimates - it would probably be a good idea to adjust some/all of them, which is what this proposal is all about.

It might be a good idea to additionally create some local type aliases for the `SmallVec`s in the `Folder` trait, as they are repeated in quite a few spots; I'd be happy to apply this sort of adjustments.

### CODE REVIEW

**Comment by llogiq** [comment]
FYI: I''ve started writing a [smallvectune](https://github.com/llogiq/smallvectune) crate that can wrap `SmallVec` and  log capacity information by internal array size over the course of the various constructors, resizings and deallocations. There are still some things to do, but once it's working, we could use this to help us gauge optimal internal array sizes for some workloads.

**Comment by nnethercote** [comment]
This PR currently makes two changes:
- remove the `OneVector` type alias (and `ExpectOne`);
- expand almost all of the formerly `OneVector` small vecs from 1 inline entry to 4 (and one in `transcribe.rs` to 8).

The first change seems fine. I agree that `OneVector` is not a particularly good name. You also mention the possibility of adding new type aliases. `SmallVec1`, `SmallVec4`, etc., would be clearer, but I'm not sure if they are necessary.

I'm less satisfied by the second change. The description above says "it is used in scenarios where the capacity of 1 is often exceeded, which might be nullifying the performance wins" -- do you have any measurements showing this? My general approach is profile-driven: I assume the existing code is reasonable, and only change such things when I have measurements showing it helps. I'm not comfortable with a blanket approach of changing every `OneVector` to have 4 inline entries without measurements. Avoiding heap allocation is good in general, but `SmallVec` can cause extra `memcpy` calls and I have seen cases where the cost of the additional copying exceeding the savings from the avoided heap allocations. It could also increase memory usage.

To move forward I would suggest removing `OneVector` and changing all its uses to `SmallVector<[T; 1]>` is fine. Beyond that, I think it's only worth changing `SmallVec` lengths when measurements show it helps. @llogiq's `smallvectune` crate could be very useful for that.

**Review by nnethercote — CHANGES_REQUESTED** [review]
(Changes requested are above.)

File: src/libsyntax/ext/tt/transcribe.rs [hunk 1]
```diff
@@ -70,7 +70,7 @@ pub fn transcribe(cx: &ExtCtxt,
                   interp: Option<FxHashMap<Ident, Rc<NamedMatch>>>,
                   src: Vec<quoted::TokenTree>)
                   -> TokenStream {
-    let mut stack: OneVector<Frame> = smallvec![Frame::new(src)];
+    let mut stack: SmallVec<[Frame; 8]> = smallvec![Frame::new(src)];
```
Comment (nnethercote) [inline src/libsyntax/ext/tt/transcribe.rs]: This is the one that is changed to 8, BTW.

**Review by nnethercote — CHANGES_REQUESTED** [review]

**Comment by ljedrz** [comment]
@nnethercote Thanks for the review! I wasn't aware these values were profiled and I now understand that changing them in an ad-hoc manner could be just as bad as spilling to the heap.

I agree that waiting for a more precise solution is the right choice. The adjusted PR should make this future change easier.

**Comment by nnethercote** [comment]
> I wasn't aware these values were profiled and I now understand that changing them in an ad-hoc manner could be just as bad as spilling to the heap.

Well, it's possible they are profiled, I don't know for sure. You could work it out with `git blame` if you really wanted. It's also possible that the exact size doesn't matter for a lot of them. But in the absence of measurements, I think it's best to leave things as they are. This is basically [Chesterton's Fence](https://en.wikipedia.org/wiki/Wikipedia:Chesterton%27s_fence) in action:

> There exists in such a case a certain institution or law; let us say, for the sake of simplicity, a fence or gate erected across a road. The more modern type of reformer goes gaily up to it and says, "I don't see the use of this; let us clear it away." To which the more intelligent type of reformer will do well to answer: "If you don't see the use of it, I certainly won't let you clear it away. Go away and think. Then, when you can come back and tell me that you do see the use of it, I may allow you to destroy it."

In this case measurements are the equivalent of the "Go away and think" part.

**Comment by llogiq** [comment]
[smallvectune](https.//crates.io/crates/smallvectune) has reached its first milestone: The data collector should work now, and uses an MPSC to collect the traces, so it's all thread-safe and shouldn't be too much of a performance burden (that said, I haven't benchmarked it...).

**Comment by llogiq** [comment]
[smallvectune](https.//crates.io/crates/smallvectune) has reached its first milestone: The data collector should work now, and uses an MPSC to collect the traces, so it's all thread-safe and shouldn't be too much of a performance burden (that said, I haven't benchmarked it...).

**Comment by TimNN** [comment]
Ping from triage! @nnethercote / @ljedrz: Could you confirm the status of this PR? IIUC, it now only removes a type alias and uses a SmallVec implementation from an external crate.

**Comment by nnethercote** [comment]
As per https://github.com/rust-lang/rust/pull/53824#issuecomment-420464001, I suggest that @ljedrz changes this PR to just be about removing `OneVector`, and then does size tuning in a follow-up.

**Comment by ljedrz** [comment]
@nnethercote actually I already did that. I'll rebase it, this change should make future size tuning easier.

**Comment by ljedrz** [comment]
Rebased.

**Comment by michaelwoerister** [comment]
Looking good!
@bors r+

**Comment by gnzlbg** [comment]
This broke interpolate idents.

**Pull request merged**